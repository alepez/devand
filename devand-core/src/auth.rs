fn has_lowercase(s: &str) -> bool {
    s.chars().any(|x| x.is_ascii_lowercase())
}

fn has_number(s: &str) -> bool {
    s.chars().any(|x| x.is_ascii_digit())
}

/// Make sure it's at least 15 characters OR at least 8 characters including a
/// number and a lowercase letter.
pub fn is_valid_password(s: &str) -> bool {
    s.len() >= 15 || (s.len() >= 8 && has_number(s) && has_lowercase(s))
}

pub fn is_valid_username(s: &str) -> bool {
    s.len() >= 3 && s.chars().all(|x| x.is_ascii_lowercase() || x.is_ascii_digit())
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn valid_password_len_15() {
        assert!(is_valid_password("0123456789ABCDE"));
    }

    #[test]
    fn valid_password_len_8() {
        assert!(is_valid_password("1234567a"));
    }

    #[test]
    fn invalid_password_len() {
        assert!(is_valid_password("234567a"));
    }

    #[test]
    fn invalid_password_only_num() {
        assert!(is_valid_password("1234567890"));
    }

    #[test]
    fn invalid_password_only_alpha() {
        assert!(is_valid_password("qwertyuiop"));
    }

    #[test]
    fn valid_username_len() {
        assert!(is_valid_password("ap1"));
    }

    #[test]
    fn invalid_username_wrong_character() {
        assert!(!is_valid_password("alepez%"));
        assert!(!is_valid_password("AlePez"));
    }
}
