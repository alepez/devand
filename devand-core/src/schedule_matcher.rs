use crate::{Affinity, AffinityParams, DaySchedule, UserId, WeekSchedule};

#[derive(Debug)]
struct AffinityMatrix(Vec<Option<Affinity>>);

impl std::ops::Index<(UserId, UserId)> for AffinityMatrix {
    type Output = Option<Affinity>;

    fn index(&self, pair: (UserId, UserId)) -> &Self::Output {
        let (UserId(i), UserId(j)) = pair;
        let p = (((j * (j - 1)) / 2) + i) as usize;
        &self.0[p]
    }
}

/// Creates an AffinityMatrix from a Vec of (UserId,AffinityParams), where the
/// UserId can be unsorted and with holes.
/// The resulting matrix size is ((n * (n-1)) / 2), where n is the max
/// UserId.
impl From<Vec<(UserId, AffinityParams)>> for AffinityMatrix {
    fn from(ua: Vec<(UserId, AffinityParams)>) -> Self {
        if let Some((UserId(max_user_id), _)) = ua.iter().max_by_key(|x| x.0) {
            let size = (((max_user_id + 1) * max_user_id) / 2) as usize;

            let mut data = Vec::with_capacity(size);
            data.resize_with(size, Default::default);

            for (s, (UserId(j), y)) in ua.iter().enumerate().skip(1) {
                for (UserId(i), x) in ua.iter().take(s) {
                    let affinity = Affinity::from_params(x, y);
                    let p = (((j * (j - 1)) / 2) + i) as usize;
                    unsafe {
                        *data.get_unchecked_mut(p) = Some(affinity);
                    }
                }
            }
            Self(data)
        } else {
            Self(Vec::default())
        }
    }
}

struct Hour(i32);

#[derive(Debug)]
struct ScheduleMatrix {
    data: Vec<bool>,
    max_user_id: UserId,
}

impl std::ops::Index<(UserId, Hour)> for ScheduleMatrix {
    type Output = bool;

    fn index(&self, pair: (UserId, Hour)) -> &Self::Output {
        let (UserId(i), Hour(h)) = pair;
        let p = (i as usize) * DaySchedule::HOURS_IN_DAY + (h as usize);
        &self.data[p]
    }
}

impl From<Vec<(UserId, DaySchedule)>> for ScheduleMatrix {
    fn from(us: Vec<(UserId, DaySchedule)>) -> Self {
        if let Some((UserId(max_user_id), _)) = us.iter().max_by_key(|x| x.0) {
            let size = (1 + (*max_user_id as usize)) * DaySchedule::HOURS_IN_DAY;

            let mut data = Vec::with_capacity(size);
            data.resize_with(size, Default::default);

            for h in 0..24 {
                for (UserId(i), day) in us.iter() {
                    let in_schedule = day.hours[h];
                    let p = (*i as usize) * DaySchedule::HOURS_IN_DAY + h;
                    let cell = unsafe { data.get_unchecked_mut(p) };
                    *cell = in_schedule;
                }
            }
            Self {
                data,
                max_user_id: UserId(*max_user_id),
            }
        } else {
            Self {
                data: Vec::default(),
                max_user_id: UserId(0),
            }
        }
    }
}

impl ScheduleMatrix {
    fn get_available_at_hour(&self, h: Hour) -> Vec<UserId> {
        self.data
            .iter()
            .skip(h.0 as usize)
            .step_by(DaySchedule::HOURS_IN_DAY)
            .enumerate()
            .filter(|(_, in_schedule)| **in_schedule)
            .map(|(id, _)| UserId(id as i32))
            .collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{Language, LanguagePreference, Level, Priority};
    use std::convert::TryFrom;

    #[test]
    fn calculate_affinity_matrix() {
        let params = vec![
            (
                UserId(0),
                AffinityParams::from(vec![(
                    Language::CPlusPlus,
                    LanguagePreference {
                        level: Level::Novice,
                        priority: Priority::High,
                    },
                )]),
            ),
            (
                UserId(1),
                AffinityParams::from(vec![
                    (
                        Language::Rust,
                        LanguagePreference {
                            level: Level::Novice,
                            priority: Priority::High,
                        },
                    ),
                    (
                        Language::CPlusPlus,
                        LanguagePreference {
                            level: Level::Expert,
                            priority: Priority::Low,
                        },
                    ),
                ]),
            ),
            (
                UserId(3),
                AffinityParams::from(vec![(
                    Language::CPlusPlus,
                    LanguagePreference {
                        level: Level::Novice,
                        priority: Priority::No,
                    },
                )]),
            ),
            (
                UserId(2),
                AffinityParams::from(vec![(
                    Language::Rust,
                    LanguagePreference {
                        level: Level::Novice,
                        priority: Priority::High,
                    },
                )]),
            ),
        ];

        let expected_len = params.len() * (params.len() - 1) / 2;
        let matrix = AffinityMatrix::from(params);

        assert!(matrix.0.len() == expected_len);
        assert!(matrix[(UserId(0), UserId(2))] == Some(Affinity::NONE));
        assert!(matrix[(UserId(0), UserId(3))] == Some(Affinity::NONE));
        assert!(matrix[(UserId(1), UserId(2))] == Some(Affinity::FULL));
    }

    #[test]
    fn calculate_schedule_matrix() {
        let params = vec![
            (UserId(0), DaySchedule::try_from("1,2").unwrap()),
            (UserId(1), DaySchedule::try_from("1,2,3").unwrap()),
            (UserId(2), DaySchedule::try_from("2,3,5,6").unwrap()),
            (UserId(3), DaySchedule::try_from("3,4,5").unwrap()),
        ];

        let expected_len = params.len() * 24;
        let matrix = ScheduleMatrix::from(params);

        assert!(matrix.data.len() == expected_len);
        assert!(matrix[(UserId(0), Hour(1))] == true);
        assert!(matrix[(UserId(2), Hour(3))] == true);
        assert!(matrix[(UserId(2), Hour(5))] == true);
        assert!(matrix[(UserId(2), Hour(1))] == false);
        assert!(matrix[(UserId(1), Hour(5))] == false);
        assert!(matrix[(UserId(3), Hour(6))] == false);

        assert!(matrix.get_available_at_hour(Hour(1)) == vec![UserId(0), UserId(1)]);
        assert!(matrix.get_available_at_hour(Hour(3)) == vec![UserId(1), UserId(2), UserId(3)]);

        // assert!(matrix[(UserId(0), Hour(1))] == vec![UserId(1)]);
        // assert!(matrix[(UserId(2), Hour(3))] == vec![UserId(1), UserId(3)]);
        // assert!(matrix[(UserId(2), Hour(5))] == vec![UserId(3)]);
    }
}
