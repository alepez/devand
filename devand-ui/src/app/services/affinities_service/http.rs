use super::FetchCallback;
use devand_core::UserAffinity;
use yew::format::{Json, Nothing};
use yew::prelude::*;
use yew::services::fetch::{FetchService, FetchTask, Request, Response};

const API_URL: &str = "/api/affinities";

pub struct AffinitiesService {
    get_handler: GetHandler,
}

struct GetHandler {
    callback: FetchCallback,
    task: Option<FetchTask>,
}

fn request<R>(
    callback: Callback<Result<Vec<UserAffinity>, anyhow::Error>>,
    r: http::request::Request<R>,
) -> Result<FetchTask, anyhow::Error>
where
    R: std::convert::Into<std::result::Result<std::string::String, anyhow::Error>>,
{
    let handler = move |response: Response<Json<Result<Vec<UserAffinity>, anyhow::Error>>>| {
        let (meta, Json(data)) = response.into_parts();
        if meta.status.is_success() {
            callback.emit(data)
        } else {
            callback.emit(Err(anyhow::anyhow!(meta.status)))
        }
    };

    FetchService::fetch(r, handler.into())
}

impl GetHandler {
    fn get(&mut self) {
        let req = Request::get(API_URL).body(Nothing).unwrap();
        self.task = request(self.callback.clone(), req).ok();
    }
}

impl AffinitiesService {
    pub fn new(callback: FetchCallback) -> Self {
        let get_handler = GetHandler {
            callback,
            task: None,
        };

        Self { get_handler }
    }

    pub fn restore(&mut self) {
        self.get_handler.get();
    }
}
