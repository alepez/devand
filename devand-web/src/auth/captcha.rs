use captcha::{CaptchaName, Difficulty};
use std::path::PathBuf;
use uuid::Uuid;

pub struct CaptchaFile {
    path: Box<PathBuf>,
    value: String,
}

impl CaptchaFile {
    pub fn value(&self) -> String {
        self.value.clone()
    }

    pub fn into_data(self) -> Vec<u8> {
        std::fs::read(self.path.as_ref()).unwrap()
    }
}

impl Drop for CaptchaFile {
    fn drop(&mut self) {
        std::fs::remove_file(self.path.as_ref()).unwrap();
    }
}

impl CaptchaFile {
    pub fn new() -> Self {
        let uuid = Uuid::new_v4();
        let filename = uuid.to_string() + ".png";
        let path = PathBuf::from("/tmp").join(filename);
        let path = Box::new(path);

        let c = captcha::by_name(Difficulty::Medium, CaptchaName::Mila);

        c.save(&path).expect("save failed");

        let value = c.chars_as_string();

        CaptchaFile { path, value }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn generate_and_destroy() {
        #[allow(unused_assignments)]
        let mut path = Box::new(PathBuf::default());
        {
            let x = CaptchaFile::new();
            path = x.path.clone();
            assert!(path.exists());
        }
        assert!(!path.exists());
    }

    #[test]
    fn captcha_value() {
        let x = CaptchaFile::new();
        let _value = x.value();
    }
}
