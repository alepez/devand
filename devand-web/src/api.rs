use crate::auth::{AuthData, LoggedUser};
use crate::{CodeNowUsers, Mailer, PgDevandConn, WeekScheduleMatrix};
use chrono::prelude::*;
use chrono::Duration;
use devand_core::schedule_matcher::AvailabilityMatch;
use devand_core::{User, UserAffinity, UserId};
use rocket::http::Status;
use rocket::{Route, State};
use rocket_contrib::json::Json;

const BASE_URL: Option<&'static str> = option_env!("DEVAND_BASE_URL");
const DEFAULT_BASE_URL: &'static str = "http://localhost:8000";

pub fn routes() -> Vec<Route> {
    routes![
        settings,
        settings_put,
        verify_email,
        affinities,
        code_now,
        availability_match,
        chat_messages_get,
        chat_messages_post,
        chat_messages_poll,
        user_public_profile,
        user_public_profile_by_id,
        password_edit,
        password_check,
    ]
}

/// Retrieve user settings
#[get("/settings")]
fn settings(user: LoggedUser) -> Json<User> {
    Json(user.into())
}

/// Update user settings
#[put("/settings", data = "<user>")]
fn settings_put(
    auth_data: AuthData,
    user: Json<User>,
    conn: PgDevandConn,
    wsmc: State<WeekScheduleMatrix>,
) -> Option<Json<User>> {
    // Note: here we don't need LoggedUser (needs db access) but only
    // auth_data to check if we are modifiyng the right user.
    if !auth_data.matches_user(&user) {
        return None;
    }

    // Update immediately the week schedule matrix
    wsmc.0
        .write()
        .unwrap()
        .update(user.id, &user.settings.schedule);

    // Save new settings in db
    devand_db::save_user(user.0, &conn.0).map(|x| Json(x))
}

/// Send a verification email to the logged user
#[post("/verify_email")]
fn verify_email(user: LoggedUser, mailer: State<Mailer>) -> Json<()> {
    mailer.verify_address(user.email.clone());
    Json(())
}

/// Retrieve user's affinities
#[get("/affinities")]
fn affinities(user: LoggedUser, conn: PgDevandConn) -> Option<Json<Vec<UserAffinity>>> {
    let users = devand_db::load_users(&conn.0)?;
    let affinities = devand_core::calculate_affinities(&user.into(), users);
    Some(Json(affinities.collect()))
}

/// Retrieve user's affinities who are online. When an user access this
/// endpoint, it is considered online for some time (see CodeNowUserMap::TTL)
#[get("/code-now")]
fn code_now(user: LoggedUser, code_now_users: State<CodeNowUsers>) -> Json<devand_core::CodeNow> {
    let user: User = user.into();

    let all_users: devand_core::CodeNowUsers = {
        // We always need a write lock, because we are updating cache ttl
        let mut cache = code_now_users.0.write().unwrap();
        cache.touch(user.clone());
        cache.clone().into()
    };

    Json(devand_core::CodeNow {
        current_user: user,
        all_users: all_users.0,
    })
}

/// Retrieve possible matching for the next week, considered user's schedule
/// and affinities
#[get("/availability-match")]
fn availability_match(user: LoggedUser, wsm: State<WeekScheduleMatrix>) -> Json<AvailabilityMatch> {
    let now = Utc::now();
    let next_week = now.checked_add_signed(Duration::days(7)).unwrap();
    let User { settings, id, .. } = user.into();
    let availability = settings.schedule;
    let wsm = wsm.0.read().unwrap();
    let wsm = wsm.get();
    let res = wsm.find_all_users_matching_in_week(id, next_week, availability);
    Json(res)
}

/// Retrieve all messages in a chat, given its members
#[get("/chat/<members>/messages")]
fn chat_messages_get(
    user: LoggedUser,
    members: String,
    conn: PgDevandConn,
) -> Option<Json<Vec<devand_core::chat::ChatMessage>>> {
    let members = parse_members(&members);

    // TODO [refactoring] Authorize using request guard
    let authorized = members.contains(&user.id);
    if !authorized {
        return None;
    }

    let result = devand_db::load_chat_history_by_members(&members, &conn);
    Some(Json(result))
}

/// Post a new message in a chat
#[post("/chat/<members>/messages", data = "<txt>")]
fn chat_messages_post(
    user: LoggedUser,
    members: String,
    txt: Json<String>,
    mailer: State<Mailer>,
    conn: PgDevandConn,
) -> Option<Json<Vec<devand_core::chat::ChatMessage>>> {
    let new_message = devand_core::chat::ChatMessage {
        author: user.id,
        txt: txt.0,
        created_at: Utc::now(),
    };

    let members = parse_members(&members);

    // TODO [refactoring] Authorize using request guard
    let authorized = members.contains(&user.id);
    if !authorized {
        return None;
    }

    crate::notifications::notify_chat_members(
        BASE_URL.unwrap_or(DEFAULT_BASE_URL),
        &mailer,
        &conn,
        &user,
        &members,
    );

    if let Some(new_message) = devand_db::add_chat_message_by_members(&members, new_message, &conn)
    {
        Some(Json(vec![new_message]))
    } else {
        None
    }
}

/// Retrieve new messages
#[get("/chat/<members>/messages/poll/<after>")]
fn chat_messages_poll(
    user: LoggedUser,
    members: String,
    after: i64,
    conn: PgDevandConn,
) -> Option<Json<Vec<devand_core::chat::ChatMessage>>> {
    // Note: Rocket 0.16.2 does not support websocket, so we just poll for new messages
    let members = parse_members(&members);

    // TODO [refactoring] Authorize using request guard
    let authorized = members.contains(&user.id);
    if !authorized {
        return None;
    }

    // TODO [optimization] It could be better loading from db only messages created after the
    // threshold, instead of filtering here.
    let result = devand_db::load_chat_history_by_members(&members, &conn)
        .into_iter()
        .filter(|x| x.created_at.timestamp() > after)
        .collect();
    Some(Json(result))
}

/// Load user public profile, given the user id. Note that this api is
/// accessible only by authenticated users, this is why we have the LoggedUser
/// guard, even if it is unused.
#[get("/u/<user_id>")]
fn user_public_profile_by_id(
    _user: LoggedUser,
    user_id: i32,
    conn: PgDevandConn,
) -> Option<Json<devand_core::PublicUserProfile>> {
    // TODO [optimization] Load only public profile
    let user = devand_db::load_user_by_id(UserId(user_id), &conn.0)?;
    Some(Json(user.into()))
}

/// Load user public profile, given the username. Note that this api is
/// accessible only by authenticated users, this is why we have the LoggedUser
/// guard, even if it is unused.
#[get("/u/<username>", rank = 2)]
fn user_public_profile(
    _user: LoggedUser,
    username: String,
    conn: PgDevandConn,
) -> Option<Json<devand_core::PublicUserProfile>> {
    // TODO [optimization] Load only public profile
    let user = devand_db::load_user_by_username(&username, &conn.0)?;
    Some(Json(user.into()))
}

/// Check if user password is the right one. This enpoint is used when
/// user is changing password.  It  just give an immediate feedback about
/// password validity.
#[post("/password-check", data = "<passwords>")]
fn password_check(
    auth_data: AuthData,
    conn: PgDevandConn,
    passwords: Json<devand_core::PasswordEdit>,
) -> Option<Json<bool>> {
    let ok = devand_db::auth::check_password(auth_data.user_id, &passwords.0.old_password, &conn)
        .ok()?;
    Some(Json(ok))
}

/// Change user password. Both old and new password must be provided, even
/// if the user is authenticated
#[post("/password-edit", data = "<passwords>")]
fn password_edit(
    auth_data: AuthData,
    conn: PgDevandConn,
    passwords: Json<devand_core::PasswordEdit>,
) -> Result<(), Status> {
    let ok = devand_db::auth::check_password(auth_data.user_id, &passwords.0.old_password, &conn)
        .map_err(|_| Status::InternalServerError)?;

    if !ok {
        return Err(Status::Unauthorized);
    }

    devand_db::auth::set_password(auth_data.user_id, &passwords.0.new_password, &conn)
        .map_err(|_| Status::InternalServerError)
}

/// Given a string with user ids separated by a dash, return a Vec of UserId
fn parse_members(s: &str) -> Vec<UserId> {
    s.split("-")
        .filter_map(|x| x.parse().ok())
        .map(UserId)
        .collect()
}
