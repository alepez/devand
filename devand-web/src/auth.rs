mod captcha;

use self::captcha::CaptchaFile;
use crate::PgDevandConn;
use core::convert::TryFrom;
use devand_db as db;
use rocket::http::{Cookie, Cookies};
use rocket::outcome::IntoOutcome;
use rocket::request::{FromRequest, Outcome, Request};
use serde::{Deserialize, Serialize};
use validator::{Validate, ValidationError};
use validator_derive::Validate;

const LOGIN_COOKIE_KEY: &'static str = "login";
const JOIN_COOKIE_KEY: &'static str = "join";
const JOIN_CAPTCHA_COOKIE_KEY: &'static str = "join_captcha";

#[derive(FromForm)]
pub struct Credentials {
    username: String,
    password: String,
}

impl Credentials {
    fn normalize(self) -> Self {
        Self {
            username: trimlow(self.username),
            password: self.password,
        }
    }
}

impl Into<db::auth::Credentials> for Credentials {
    fn into(self) -> db::auth::Credentials {
        db::auth::Credentials {
            username: self.username,
            password: self.password,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AuthData {
    pub user_id: i32,
}

impl AuthData {
    pub fn matches_user(&self, user: &devand_core::User) -> bool {
        user.id == self.user_id
    }
}

impl TryFrom<rocket::http::Cookie<'_>> for AuthData {
    type Error = ();
    fn try_from(cookie: rocket::http::Cookie<'_>) -> Result<Self, Self::Error> {
        let json = cookie.value();
        serde_json::from_str(json).or(Err(()))
    }
}

impl<'a> Into<rocket::http::Cookie<'a>> for AuthData {
    fn into(self) -> rocket::http::Cookie<'a> {
        let json = serde_json::to_string(&self).unwrap();
        Cookie::build(LOGIN_COOKIE_KEY, json)
            .http_only(true)
            .max_age(time::Duration::days(30))
            .finish()
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for AuthData {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<AuthData, ()> {
        request
            .cookies()
            .get_private(LOGIN_COOKIE_KEY)
            .and_then(|cookie| AuthData::try_from(cookie).ok())
            .or_forward(())
    }
}

pub(crate) fn login(
    cookies: &mut Cookies,
    credentials: Credentials,
    conn: &PgDevandConn,
) -> Result<(), ()> {
    let credentials = credentials.normalize().into();

    db::auth::login(credentials, &conn.0)
        .map(|user_id| {
            let auth_data = AuthData { user_id };
            cookies.add_private(auth_data.into());
        })
        .map_err(|_| ())
}

pub fn logout(cookies: &mut Cookies) {
    cookies.remove_private(Cookie::named(LOGIN_COOKIE_KEY));
}

#[derive(FromForm, Debug, Validate, Serialize, Deserialize, Clone)]
pub struct JoinData {
    #[validate(custom = "validate_username")]
    pub username: String,
    #[validate(email)]
    pub email: String,
    #[validate(custom = "validate_password")]
    pub password: String,
    pub captcha: String,
}

impl TryFrom<rocket::http::Cookie<'_>> for JoinData {
    type Error = ();
    fn try_from(cookie: rocket::http::Cookie<'_>) -> Result<Self, Self::Error> {
        let json = cookie.value();
        serde_json::from_str(json).or(Err(()))
    }
}

impl<'a> Into<rocket::http::Cookie<'a>> for JoinData {
    fn into(self) -> rocket::http::Cookie<'a> {
        let json = serde_json::to_string(&self).unwrap();
        Cookie::build(JOIN_COOKIE_KEY, json)
            .http_only(true)
            .max_age(time::Duration::minutes(10))
            .path("/join")
            .finish()
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for JoinData {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<JoinData, ()> {
        request
            .cookies()
            .get_private(JOIN_COOKIE_KEY)
            .and_then(|cookie| JoinData::try_from(cookie).ok())
            .or_forward(())
    }
}

impl JoinData {
    fn normalize(self) -> Self {
        Self {
            username: trimlow(self.username),
            email: trimlow(self.email),
            password: self.password,
            captcha: trimlow(self.captcha),
        }
    }
}

fn validate_password(s: &str) -> Result<(), validator::ValidationError> {
    if devand_core::auth::is_valid_password(s) {
        Ok(())
    } else {
        let mut err = ValidationError::new("invalid_password");
        err.message = Some("Make sure it's at least 15 characters OR at least 8 characters including a number and a lowercase letter".into());
        Err(err)
    }
}

fn validate_username(s: &str) -> Result<(), validator::ValidationError> {
    if devand_core::auth::is_valid_username(s) {
        Ok(())
    } else {
        let mut err = ValidationError::new("invalid_password");
        err.message = Some("Must be at least 3 characters, only lowercase and digits".into());
        Err(err)
    }
}

fn is_unique_email(s: &str, conn: &PgDevandConn) -> Result<(), validator::ValidationError> {
    if devand_db::is_email_available(s, conn) {
        Ok(())
    } else {
        let mut err = ValidationError::new("email_unavailable");
        err.message = Some("Email is already used".into());
        Err(err)
    }
}

fn is_unique_username(s: &str, conn: &PgDevandConn) -> Result<(), validator::ValidationError> {
    if devand_db::is_username_available(s, conn) {
        Ok(())
    } else {
        let mut err = ValidationError::new("username_unavailable");
        err.message = Some("Username is already used".into());
        Err(err)
    }
}

fn validate_captcha(input: String, expected: String) -> Result<(), validator::ValidationError> {
    let input = trimlow(input);
    let expected = trimlow(expected);

    if input == expected {
        Ok(())
    } else {
        let mut err = ValidationError::new("captcha");
        err.message = Some("Wrong captcha".into());
        Err(err)
    }
}

fn check_for_uniqueness(
    join_data: &JoinData,
    conn: &PgDevandConn,
) -> Result<(), validator::ValidationErrors> {
    let mut errors = validator::ValidationErrors::new();

    if let Err(e) = is_unique_username(&join_data.username, conn) {
        errors.add("username", e);
    }

    if let Err(e) = is_unique_email(&join_data.email, conn) {
        errors.add("email", e);
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}

pub(crate) fn join(
    cookies: &mut Cookies,
    join_data: JoinData,
    expected_captcha: ExpectedCaptcha,
    conn: &PgDevandConn,
) -> Result<(), JoinError> {
    let mut join_data = join_data.normalize();

    let valid = join_data.validate();

    let captcha_input = join_data.captcha;
    join_data.captcha = String::new();

    // Remove old cookie
    cookies.remove_private(Cookie::named(JOIN_COOKIE_KEY));
    cookies.remove_private(Cookie::named(JOIN_CAPTCHA_COOKIE_KEY));

    let valid = valid
        .and_then(|_| check_for_uniqueness(&join_data, conn))
        .and_then(|_| {
            let mut errors = validator::ValidationErrors::new();

            if let Err(e) = validate_captcha(captcha_input, expected_captcha.value) {
                errors.add("captcha", e);
            }

            if errors.is_empty() {
                Ok(())
            } else {
                Err(errors)
            }
        });

    // Add new join data if there is an error
    if valid.is_err() {
        // Update with new data
        cookies.add_private(join_data.clone().into());
    }

    // Early exit if not valid
    valid?;

    let join_data = db::auth::JoinData {
        username: join_data.username,
        password: join_data.password.into(),
        email: join_data.email,
    };

    let result = db::auth::join(join_data, &conn.0);

    if result.is_ok() {
        cookies.remove_private(Cookie::named(JOIN_COOKIE_KEY));
        cookies.remove_private(Cookie::named(JOIN_CAPTCHA_COOKIE_KEY));
    }

    result.map(|_| ()).map_err(|_| JoinError::UnknownError)
}

pub(crate) fn captcha(cookies: &mut Cookies) -> Result<CaptchaFile, ()> {
    let captcha = CaptchaFile::new();

    let captcha_value = ExpectedCaptcha {
        value: captcha.value(),
    };

    cookies.add_private(captcha_value.into());

    Ok(captcha)
}

#[derive(Debug)]
pub enum JoinError {
    UnknownError,
    ValidationError(String),
}

impl From<validator::ValidationErrors> for JoinError {
    fn from(e: validator::ValidationErrors) -> Self {
        let msg = e
            .field_errors()
            .into_iter()
            .map(|(field, errors)| {
                let msg = errors
                    .iter()
                    .filter_map(|x| x.message.clone())
                    .collect::<Vec<_>>()
                    .concat();
                if msg.is_empty() {
                    format!("Invalid {}. ", field)
                } else {
                    format!("Invalid {}: {}. ", field, msg)
                }
            })
            .collect::<Vec<_>>()
            .concat();
        JoinError::ValidationError(msg)
    }
}

impl ToString for JoinError {
    fn to_string(&self) -> String {
        match self {
            JoinError::UnknownError => "UnknownError".into(),
            JoinError::ValidationError(msg) => msg.into(),
        }
    }
}

fn trimlow(s: String) -> String {
    // We help the user, trimming spaces and converting to lowercase
    let s = s.to_lowercase();
    // Note: this allocates a new string, in place trimming does not exist
    s.trim().to_string()
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ExpectedCaptcha {
    pub value: String,
}

impl TryFrom<rocket::http::Cookie<'_>> for ExpectedCaptcha {
    type Error = ();
    fn try_from(cookie: rocket::http::Cookie<'_>) -> Result<Self, Self::Error> {
        let json = cookie.value();
        serde_json::from_str(json).or(Err(()))
    }
}

impl<'a> Into<rocket::http::Cookie<'a>> for ExpectedCaptcha {
    fn into(self) -> rocket::http::Cookie<'a> {
        let json = serde_json::to_string(&self).unwrap();
        Cookie::build(JOIN_CAPTCHA_COOKIE_KEY, json)
            .http_only(true)
            .max_age(time::Duration::minutes(10))
            .path("/join")
            .finish()
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for ExpectedCaptcha {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<ExpectedCaptcha, ()> {
        request
            .cookies()
            .get_private(JOIN_CAPTCHA_COOKIE_KEY)
            .and_then(|cookie| ExpectedCaptcha::try_from(cookie).ok())
            .or_forward(())
    }
}

pub struct RealIp(pub std::net::IpAddr);

impl<'a, 'r> FromRequest<'a, 'r> for RealIp {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<RealIp, ()> {
        let real_ip_from_header = request.real_ip();
        let remote_ip_from_sock = request.remote().map(|x| x.ip());

        real_ip_from_header
            .or(remote_ip_from_sock)
            .map(|x| RealIp(x))
            .or_forward(())
    }
}

pub struct LoggedUser(devand_core::User);

impl<'a, 'r> FromRequest<'a, 'r> for LoggedUser {
    type Error = ();

    fn from_request(request: &'a Request<'r>) -> Outcome<LoggedUser, ()> {
        let conn = request.guard::<PgDevandConn>()?;
        let auth_data = request.guard::<AuthData>()?;
        let user = devand_db::load_user_by_id(auth_data.user_id, &conn.0);
        user.map(|x| LoggedUser(x)).or_forward(())
    }
}

impl Into<devand_core::User> for LoggedUser {
    fn into(self) -> devand_core::User {
        self.0
    }
}

impl std::ops::Deref for LoggedUser {
    type Target = devand_core::User;

    #[inline(always)]
    fn deref(&self) -> &devand_core::User {
        &self.0
    }
}
