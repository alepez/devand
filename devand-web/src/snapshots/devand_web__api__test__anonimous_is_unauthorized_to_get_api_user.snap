---
source: devand-web/src/api.rs
expression: response.body_string().unwrap()
---
 <!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>DevAndDev - Unauthorized</title>
  <link rel="stylesheet" href="/static/ui-HASHPLACEHOLDER/pure.css">
  <link rel="stylesheet" href="/static/ui-HASHPLACEHOLDER/grids-responsive.css">
  <link rel="stylesheet" href="/static/ui-HASHPLACEHOLDER/main.css">
  <link rel="apple-touch-icon-precomposed" href="/static/favicon/favicon-180.png">
  <link rel="icon" href="/static/favicon/favicon-32.png" sizes="32x32">
  <link rel="icon" href="/static/favicon/favicon-128.png" sizes="128x128">
  <link rel="icon" href="/static/favicon/favicon-180.png" sizes="180x180">
  <link rel="icon" href="/static/favicon/favicon-196.png" sizes="196x196">
</head>

<body>
<div class="container">
    <header class="main-header">
    <div class="main-header-content">
      <div class="pure-menu pure-menu-horizontal">
        <ul class="pure-menu-list">
          <li class="pure-menu-item">
            <a href="/" title="DevAndDev" class="pure-menu-link"><img src="/static/favicon/favicon-128.png" /></a>
          </li>
    
          <li class="pure-menu-item">
            <a href="/login" class="pure-menu-link">Sign In</a>
          </li>
          <li class="pure-menu-item">
            <a href="/join" class="pure-menu-link">Sign Up</a>
          </li>
    
        </ul>
      </div>
    </div>
  </header>


  <div class="content">
    <h1>Unauthorized</h1>

<p>The request requires user authentication.</p>

<p>Please, <a href="/login/%2Fapi%2Fuser">sign in</a></p>


  </div>
  <footer class="main-footer">
  <div class="main-footer-content">
    
    <div class="devand-cookie-policy">
      <p>We use cookies in the delivery of our services. To learn about the cookies we use and information about your preferences and opt-out choices, please <a href="/privacy">click here</a>. By using our platform you agree to our use of cookies.</p>
    </div>
    
    <ul class="pure-g">
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="/help">Help</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="/code-of-conduct">Code of Conduct</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="/privacy">Privacy Policy</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="https://github.com/alepez/devand">Code</a>
      </li>
      <li class="pure-u-1 pure-u-md-1-3">
        <a href="https://alepez.me">Â© 2020 Alessandro Pezzato</a>
      </li>
    </ul>
  </div>
</footer>

</div>
</body>

</html>


